### DISCUSSION TITLE
Face grouping

### DISCUSSION THREAD

**Discussion by nicnet123** [exadel-inc/CompreFace#1049] [General]
Just started playing with Compreface yesterday, and am impressed. This is is a really hands on face recognition system that offers the flexibility I need with my large photo collection, unlike some of the other photo packages out there, that offer recognition, but it's tightly controlled.
One thing I would find useful if to classify faces into groups - so set a landmark threshold, and categorize similar faces together. So if I have 100k faces, group them into say 1000 (depending on the threshold) groups. 
At the moment, recognition compares landmarks to a set of trained faces.
Instead, I would like it to compare against other untrained faces.
Is possible?

**Comment by pospielov** [comment]
Not sure I totally understand your case.
But there are two ways how you can perform face grouping:
1. By applying a clustering algorithm. In this case, you use a CompreFace face detection service with a `calculator` plugin to gather embedding for each face. Then you apply locally one of the clustering algorithms, e.g., k-means or DBSCAN. Then using the results, you save the faces in CompreFace.
I recommend using this approach even though it will require some ML knowledge. 
2. By saving images one by one and using similarity and threshold. In this case, you define a threshold that says that the two faces are identical. So you:
- save a first face
- use the recognition endpoint to see the similarity with the second image
- if the similarity exceeds the threshold, you save the image to the same subject. If not, you create a new subject.
- repeat the same for each next image
The disadvantage of this approach is that it's not an easy task to choose the threshold. So if it's too small, you will merge different people into one subject; if it's too large, you will split the same person into two subjects.

**Reply by nicnet123** [reply]
thanks for you excellent answers.
1. is currently beyond my level - I started to look at embeddings and k-means, and right now I haven't the time. Also, probably overkill for my use case (I will post my workflow and a question separately(.
2. Much easier, but I can see it getting bogged down quickly, as need to compare each face to all previous saved images. Initially fine, but when there are hundreds, or thousands, it will start to take time.
I try a different approach.
